docs/ADR-10-cardano-api-script-witness-api.md

```diff
@@ -0,0 +1,271 @@
+- Use `cardano-ledger`'s `PlutusRunnable` type in order to leverage it's serialization instances that check the validity of the plutus script at the point of disk read. 
+- The new api needs to be compatible with the old api
+
+# Decision
+
+## Summary
+
+Currently we have several functions that generate indicies in order to construct the redeemer pointer map. For example:
+
+```haskell
+
+indexTxWithdrawals
+  :: TxWithdrawals BuildTx era
+  -> [(ScriptWitnessIndex, StakeAddress, L.Coin, Witness WitCtxStake era)]
+indexTxWithdrawals TxWithdrawalsNone = []
+indexTxWithdrawals (TxWithdrawals _ withdrawals) =
+  [ (ScriptWitnessIndexWithdrawal ix, addr, coin, witness)
+  | (ix, (addr, coin, BuildTxWith witness)) <- zip [0 ..] (orderStakeAddrs withdrawals)
+  ]
+
+...
+
+indexTxMintValue
+  :: TxMintValue build era
+  -> [ ( ScriptWitnessIndex
+       , PolicyId
+       , AssetName
+       , Quantity
+       , BuildTxWith build (ScriptWitness WitCtxMint era)
+       )
+     ]
+indexTxMintValue TxMintNone = []
+indexTxMintValue (TxMintValue _ policiesWithAssets) =
+  [ (ScriptWitnessIndexMint ix, policyId', assetName', quantity, witness)
+  | (ix, (policyId', assets)) <- zip [0 ..] $ toList policiesWithAssets
+  , (assetName', quantity, witness) <- assets
+  ]
+
+...
+
+indexTxIns
+  :: TxIns BuildTx era
+  -> [(ScriptWitnessIndex, TxIn, Witness WitCtxTxIn era)]
+indexTxIns txins =
+  [ (ScriptWitnessIndexTxIn ix, txIn, witness)
+  | (ix, (txIn, BuildTxWith witness)) <- zip [0 ..] $ orderTxIns txins
+  ]
+
+...etc
+
+```
+
+There is a repeating pattern here that basically has something that is _witnessable_ (tx input, mint value, certificate, etc) accompanied by its index and the corresponding witness. Therefore the following types are introduced to capture this pattern:
+
+```haskell
+data Witnessable thing era where
+  WitTxIn :: L.AlonzoEraScript era => TxIn -> Witnessable TxIn era
+  WitTxCert :: L.AlonzoEraScript era => Cert -> Witnessable Cert era
+  WitMint :: L.AlonzoEraScript era => Mint -> Witnessable Mint era
+  WitWithdrawal
+    :: L.AlonzoEraScript era => Withdrawal -> Witnessable Withdrawal era
+  WitVote
+    :: L.ConwayEraScript era
+    => Voter -> Witnessable Voter era
+  WitProposal :: L.ConwayEraScript era => Proposal -> Witnessable Proposal era
+
+instance Ord (Witnessable thing era) where
+  compare a b =
+    case (a, b) of
+      (WitTxIn txinA, WitTxIn txinB) -> compare txinA txinB
+      (WitTxCert (certA, _), WitTxCert (certB, _)) -> compare certA certB
+      (WitMint mintA, WitMint mintB) -> compare mintA mintB
+      (WitWithdrawal (stakeAddrA, _), WitWithdrawal (stakeAddrB, _)) -> compare stakeAddrA stakeAddrB
+      (WitVote voterA, WitVote voterB) -> compare voterA voterB
+      (WitProposal propA, WitProposal propB) -> compare propA propB
+
+type Mint = (PolicyId, AssetName, Quantity)
+
+type Withdrawal = (StakeAddress, L.Coin)
```

Comment:
Is the idea to use those `Mint`, `Withdrawal` ... types inside `Tx*` types, like `TxMint` etc?

Reply:
So actually this sets the stage for the successor type of `TxBodyContent`, which should be named something like `UnsignedTx`:

```
data UnsignedTx era
  = UnsignedTx
  { txIns :: [(Witnessable TxIn era, AnyWitness era)]
  , ...
  , txWithdrawals :: [(Witnessable Withdrawal era, AnyWitness era)]
  , ...
```

Reply:
We can decide if we want to modify the Tx* themselves and reuse them.